src/buffer.c

```diff
@@ -309,19 +350,44 @@ int BUFFER_append(BUFFER_HANDLE handle1, BUFFER_HANDLE handle2)
         }
         else
         {
-            unsigned char* temp = (unsigned char*)realloc(b1->buffer, b1->size + b2->size);
-            if (temp == NULL)
+            if (b2->size ==0)
             {
-                /* Codes_SRS_BUFFER_07_023: [BUFFER_append shall return a nonzero upon any error that is encountered.] */
-                result = __LINE__;
+                // b2->size = 0, whatever b1->size is, do nothing
+                result = 0;
+            }
+            else if (b1->size == 0 && b2->size !=0)
+            {
+                unsigned char* temp = (unsigned char*)malloc(b2->size);
+                if (temp == NULL)
+                {
+                    /* Codes_SRS_BUFFER_07_023: [BUFFER_append shall return a nonzero upon any error that is encountered.] */
+                    result = __LINE__;
+                }
+                else
+                {
+                    b1->buffer = temp;
+                    memcpy(b1->buffer, b2->buffer, b2->size);
+                    b1->size = b2->size;
+                    result = 0;
+                }
             }
```

Comment:
I agree the size of the allocated memory for b1->buffer is 1 byte.
b1->size is 0.
b2->size is let's say 10.

The amount of memory to be allocated by realloc should be 10 bytes (b1->size + b2->size). Forget about the fact that the extra byte was allocated for b1 ...

Thus there is no need for special handling.

---
In reply to: [75432767](https://github.com/Azure/azure-c-shared-utility/pull/20#discussion_r75432767) [](ancestors = 75432767)